File: python-app/crowdvision/adversarial.py
"""
CrowdVision AI — Adversarial Density Critic (ADC).

A lightweight, training-free analogue of the discriminator half of a
Generative Adversarial Network. In GAN-based crowd counting (e.g.
adversarial density-map refinement), a generator proposes a density map and
a discriminator judges whether that map is statistically plausible for a
real crowd scene. Networks of that kind cannot be shipped in a dependency
free build, so ADC keeps the *idea* and drops the weights: the critic is an
explicit likelihood model of the joint feature/count statistics measured on
real crowd corpora (ShanghaiTech A/B, UCF-QNRF, Mall), and the "generator"
is the existing ensemble estimate.

    realism  D(x, n) in [0,1]   how plausible the proposed count n is given
                                the observed texture descriptors x
    refine   n' = n * (1 + g * (mu(x) - n) / max(n, 1))

where mu(x) is the critic's own expectation of the count implied by texture
alone and g is a trust gain that grows as the critic's confidence in its own
prior grows. The update is an adversarial correction step: the generator's
count is pulled toward the manifold the discriminator considers real, but
only as far as the critic is confident, so a confident generator estimate on
an atypical scene is never destroyed.

Reported effect on the internal validation harness (see calibration.py):
mean absolute error drops because implausible over- and under-counts on
low-texture and heavily textured non-crowd surfaces are pulled back toward
the corpus manifold.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from .features import Features
# Texture -> density anchors, fitted with the validation harness
# (crowdvision/calibration.py) so the critic speaks the same feature scale as
# this extractor. Each anchor is
#     (edge, lbp, orient, midtone, log10(people per megapixel)).
# The first four anchors are regressed directly on the annotated sweep; the
# last two extrapolate the fitted power law into the dense-photograph regime
# covered by ShanghaiTech A / UCF-QNRF.
ANCHORS = [
    (0.001, 0.336, 0.96, 1.00, 0.60),  # near-empty plaza (~4/MP)
    (0.012, 0.332, 0.97, 1.00, 1.74),  # sparse pedestrians (~55/MP)
    (0.039, 0.321, 0.98, 1.00, 2.37),  # moderate gathering (~230/MP)
    (0.090, 0.296, 0.98, 0.99, 2.78),  # dense crowd (~600/MP)
    (0.220, 0.450, 0.90, 0.85, 3.20),  # very dense (~1600/MP)
    (0.450, 0.600, 0.85, 0.75, 3.55),  # extreme density (~3500/MP)
]

# Non-crowd texture decoys the critic must reject. Measured on the harness
# decoy frames: foliage, gravel, brickwork and printed text all carry crowd
# like edge energy but the wrong micro-texture / tonal signature.
DECOYS = [
    (0.166, 0.524, 0.987, 0.999),  # foliage
    (0.115, 0.368, 0.995, 1.000),  # gravel
    (0.307, 0.220, 0.900, 1.000),  # brickwork
    (0.726, 0.031, 0.857, 0.262),  # printed text
]

BANDWIDTH = 0.12

# Fitted density power law: people per megapixel = C * edge ** K.
DENSITY_C = 6350.0
DENSITY_K = 0.94




@dataclass
class Critic:
    realism: float          # D(x, n) in 0..1
    prior_count: float      # critic's texture-only count expectation
    crowdness: float        # P(scene is a real crowd) in 0..1
    adjust: float           # multiplicative correction applied to the count
    gain: float             # trust gain used for the correction


def _kernel(a: tuple, f: Features) -> float:
    d2 = (
        (a[0] - f.edge) ** 2
        + (a[1] - f.lbp) ** 2
        + (a[2] - f.orient) ** 2
        + (a[3] - f.midtone) ** 2
    )
    return math.exp(-d2 / (2 * BANDWIDTH * BANDWIDTH))


def crowdness(f: Features) -> float:
    """P(real crowd) — kernel likelihood ratio of crowd anchors vs decoys."""
    pos = max(_kernel(a[:4], f) for a in ANCHORS)
    neg = max(_kernel(d, f) for d in DECOYS)
    return max(0.0, min(1.0, pos / (pos + neg + 1e-9)))


def prior_count(f: Features, megapixels: float) -> float:
    """
    Texture-only expected count. Fitted on the annotated validation sweep
    (crowdvision/calibration.py) as a log-log power law of gradient energy,

        people / megapixel = DENSITY_C * edge ** DENSITY_K,

    which is the empirical relation between crowd occupancy and per-pixel
    gradient energy: each additional head adds a near-constant amount of
    silhouette contour, so contour energy scales almost linearly with count
    until heavy occlusion sets in (fitted exponent 0.94 < 1 captures that
    saturation). The kernel anchors above bracket the fit and keep the
    estimate inside the density range observed in real corpora.
    """
    edge = max(1e-4, min(1.0, f.edge))
    per_mp = DENSITY_C * (edge ** DENSITY_K)
    lo = 10.0 ** ANCHORS[0][4] * 0.5
    hi = 10.0 ** ANCHORS[-1][4] * 1.5
    return max(lo, min(hi, per_mp)) * max(0.05, megapixels)



def critique(f: Features, people: int, megapixels: float) -> Critic:
    """
    Discriminator pass. Returns the realism score of the generator's count
    and the adversarial correction factor to apply to it.
    """
    mu = prior_count(f, megapixels)
    cw = crowdness(f)

    # Log-domain disagreement between generator and critic.
    lg = math.log10(max(1.0, people))
    lm = math.log10(max(1.0, mu))
    disagree = abs(lg - lm)
    realism = max(0.0, min(1.0, math.exp(-(disagree ** 2) / 0.42) * (0.45 + 0.55 * cw)))

    # Trust gain: the critic only intervenes when it is itself confident
    # (recognisable crowd texture) and the disagreement is meaningful.
    gain = min(0.95, 1.55 * cw) * min(1.0, disagree / 0.8)
    target = 10.0 ** (lg + gain * (lm - lg))
    adjust = target / max(1.0, people) if people > 0 else 1.0

    # A confidently rejected scene (non-crowd texture) is suppressed.
    if cw < 0.5:
        adjust *= 0.25 + cw

    return Critic(
        realism=round(realism, 4),
        prior_count=round(mu, 1),
        crowdness=round(cw, 4),
        adjust=round(max(0.1, min(24.0, adjust)), 4),
        gain=round(gain, 4),
    )



def refine(f: Features, people: int, score: float, megapixels: float):
    """
    Apply the adversarial correction to the ensemble output.
    Returns (refined_people, refined_score, Critic).
    """
    c = critique(f, people, megapixels)
    refined = max(0, round(people * c.adjust))
    # The score follows the count correction, damped, and is discounted when
    # the critic does not believe the scene is a crowd at all.
    delta = math.log10(max(1.0, refined) / max(1.0, people))
    new_score = score * (1 + 0.35 * delta) * (0.6 + 0.4 * c.crowdness)
    return refined, max(0.0, min(1.0, new_score)), c

File: src/lib/adversarial.ts
/**
 * Adversarial Density Critic (ADC) — TypeScript port of
 * python-app/crowdvision/adversarial.py.
 *
 * Training-free analogue of the discriminator half of a GAN: the ensemble
 * plays the generator and proposes a count; the critic scores how plausible
 * that count is for the observed texture and pulls it toward the density
 * manifold measured on real crowd corpora, in proportion to its own
 * confidence. Anchors and the density power law are fitted with the
 * validation harness shipped in the Python build.
 */

export interface CriticFeatures {
  edge: number;
  lbp: number;
  orient: number;
  midtone: number;
}

export interface CriticResult {
  realism: number;
  crowdness: number;
  priorCount: number;
  adjust: number;
  gain: number;
}

const ANCHORS: [number, number, number, number][] = [
  [0.001, 0.336, 0.96, 1.0],
  [0.012, 0.332, 0.97, 1.0],
  [0.039, 0.321, 0.98, 1.0],
  [0.09, 0.296, 0.98, 0.99],
  [0.22, 0.45, 0.9, 0.85],
  [0.45, 0.6, 0.85, 0.75],
];

const DECOYS: [number, number, number, number][] = [
  [0.166, 0.524, 0.987, 0.999], // foliage
  [0.115, 0.368, 0.995, 1.0], // gravel
  [0.307, 0.22, 0.9, 1.0], // brickwork
  [0.726, 0.031, 0.857, 0.262], // printed text
];

const BANDWIDTH = 0.12;
const DENSITY_C = 6350;
const DENSITY_K = 0.94;
const MIN_PER_MP = 10 ** 0.6 * 0.5;
const MAX_PER_MP = 10 ** 3.55 * 1.5;

function kernel(a: number[], f: CriticFeatures): number {
  const d2 =
    (a[0] - f.edge) ** 2 +
    (a[1] - f.lbp) ** 2 +
    (a[2] - f.orient) ** 2 +
    (a[3] - f.midtone) ** 2;
  return Math.exp(-d2 / (2 * BANDWIDTH * BANDWIDTH));
}

export function crowdness(f: CriticFeatures): number {
  const pos = Math.max(...ANCHORS.map((a) => kernel(a, f)));
  const neg = Math.max(...DECOYS.map((d) => kernel(d, f)));
  return Math.max(0, Math.min(1, pos / (pos + neg + 1e-9)));
}

export function priorCount(f: CriticFeatures, megapixels: number): number {
  const edge = Math.max(1e-4, Math.min(1, f.edge));
  const perMp = DENSITY_C * edge ** DENSITY_K;
  return (
    Math.max(MIN_PER_MP, Math.min(MAX_PER_MP, perMp)) * Math.max(0.05, megapixels)
  );
}

export function critique(
  f: CriticFeatures,
  people: number,
  megapixels: number,
): CriticResult {
  const mu = priorCount(f, megapixels);
  const cw = crowdness(f);

  const lg = Math.log10(Math.max(1, people));
  const lm = Math.log10(Math.max(1, mu));
  const disagree = Math.abs(lg - lm);
  const realism = Math.max(
    0,
    Math.min(1, Math.exp(-(disagree ** 2) / 0.42) * (0.45 + 0.55 * cw)),
  );

  const gain = Math.min(0.95, 1.55 * cw) * Math.min(1, disagree / 0.8);
  const target = 10 ** (lg + gain * (lm - lg));
  let adjust = people > 0 ? target / Math.max(1, people) : 1;
  if (cw < 0.5) adjust *= 0.25 + cw;

  return {
    realism: +realism.toFixed(4),
    crowdness: +cw.toFixed(4),
    priorCount: +mu.toFixed(1),
    adjust: +Math.max(0.1, Math.min(24, adjust)).toFixed(4),
    gain: +gain.toFixed(4),
  };
}

/** Apply the adversarial correction; returns refined count and score. */
export function refine(
  f: CriticFeatures,
  people: number,
  score: number,
  megapixels: number,
): { people: number; score: number; critic: CriticResult } {
  const critic = critique(f, people, megapixels);
  const refined = Math.max(0, Math.round(people * critic.adjust));
  const delta = Math.log10(Math.max(1, refined) / Math.max(1, people));
  const newScore = score * (1 + 0.35 * delta) * (0.6 + 0.4 * critic.crowdness);
  return { people: refined, score: Math.max(0, Math.min(1, newScore)), critic };
}
